libcli.c

```diff
@@ -1362,49 +1301,134 @@ int cli_loop(struct cli_def *cli, int sockfd) {
+            _write(sockfd, comphelp.entries[i], strlen(comphelp.entries[i]));
           }
-          if (i % 4 != 3) _write(sockfd, "\r\n", 2);
+          _write(sockfd, "\r\n", 2);
           cli->showprompt = 1;
-        } else {
-          // More than one completion
+        } else if (comphelp.num_entries == 1) {
+          // Single completion - show *unless* the optional/required 'prefix' is present
+          if ((comphelp.entries[0][0] != '[') && (comphelp.entries[0][0] != '<')) {
+            for (; l > 0; l--, cursor--) {
+              if (cmd[l - 1] == ' ' || cmd[l - 1] == '|' || (comphelp.comma_separated && cmd[l - 1] == ',')) break;
+              _write(sockfd, "\b", 1);
+            }
+            strcpy((cmd + l), comphelp.entries[0]);
+            l += strlen(comphelp.entries[0]);
+            cmd[l++] = ' ';
+            cursor = l;
+            _write(sockfd, comphelp.entries[0], strlen(comphelp.entries[0]));
+            _write(sockfd, " ", 1);
+            // and now forget the tab, since we just found a single match
+            lastchar = '\0';
+          } else {
+            // Yes, we had a match, but it wasn't required - remember the tab in case the user double tabs....
+            lastchar = CTRL('I');
+          }
+        } else if (comphelp.num_entries > 1) {
+          /* More than one completion
+           * Show as many characters as we can until the completions start to differ
+           */
           lastchar = c;
-          _write(sockfd, "\a", 1);
+          int i, j, k = 0;
+          char *tptr = comphelp.entries[0];
+
+          /* quickly try to see where our entries differ
+           * corner cases
+           * - if all entries are optional, don't show
+           *   *any* options unless user has provided a letter.
+           * - if any entry starts with '<' then don't fill in
+           *   anything.
+           */
+
+          // skip a leading '['
+          k = strlen(tptr);
+          if (*tptr == '[')
+            tptr++;
+          else if (*tptr == '<')
+            k = 0;
+
+          for (i = 1; k != 0 && i < comphelp.num_entries; i++) {
+            char *wptr = comphelp.entries[i];
+
+            if (*wptr == '[')
+              wptr++;
+            else if (*wptr == '<')
+              k = 0;
+
+            for (j = 0; (j < k) && (j < (int)strlen(wptr)); j++) {
+              if (strncasecmp(tptr + j, wptr + j, 1)) break;
+            }
+            k = j;
+          }
+
+          /*
+           * ok, try to show minimum match string if we have a
+           * non-zero k and the first letter of the last word
+           * is not '['
+           */
+          if (k && (comphelp.entries[comphelp.num_entries - 1][0] != '[')) {
+            for (; l > 0; l--, cursor--) {
+              if (cmd[l - 1] == ' ' || cmd[l - 1] == '|' || (comphelp.comma_separated && cmd[l - 1] == ',')) break;
+              _write(sockfd, "\b", 1);
+            }
+            strncpy((cmd + l), tptr, k);
```

Comment:
I'm a fan of less parentheses when they aren't required
```c
strncpy(cmd + 1, tptr, k);
```

Reply:
Won't disagree - fixed.